src/Microsoft.Extensions.Logging.Abstractions/ISupportsExternalScope.cs

```diff
@@ -0,0 +1,21 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.Logging
+{
+    /// <summary>
+    /// An interface for configuring logging providers.
+    /// </summary>
+    public interface ISupportsExternalScope
+    {
+        void SetScopeProvider(IExternalScopeProvider scopeProvider);
+    }
+
+    public interface IExternalScopeProvider
+    {
+        void CollectScope<T>(Action<object, T> callback, T state);
```

Comment:
Nit: I prefer AggregateScopes or ReduceScopes.

Maybe `TAccumulate AggregateScopes<TAccumulate>(Function<TAccumulate, object, TAccumulate>`. It should be easy to convert any `Action<object, T> callback` into this by just returning the T/TAccumlate.

I actually don't care to much if the method is void or not, but putting Aggregate or Reduce in the name makes it more clear that the callback is called potentially multiple times inline.

Reply:
`AggregateScopes` is fine with me if it lets you flow an accumulator, to match `Enumerable.Aggregate`. I'm not a fan of `ReduceScopes` because `Reduce` isn't a term we generally use in our APIs (even if I do like it :)).

I agree `Collect` isn't a great word since it implies collecting into a single unit (kinda like `Aggregate`). Perhaps just `ForEachScope` (precedent with `List<T>.ForEach`) or `EnumerateScopes` (other BCL examples of this return `IEnumerable` but I think we wouldn't want to do that here, however it would be pretty clear what the callback is doing when you see the `void` return).

Reply:
I like `ForEach` more, don't think `Aggregate` pattern would be useful in this case: `TAccumulate` would often be collection or collection like object and passed in by reference.